src/compiler/checker.ts

```diff
@@ -13465,26 +13470,70 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getPropertiesOfUnionOrIntersectionType(type: UnionOrIntersectionType): Symbol[] {
-        if (!type.resolvedProperties) {
-            const members = createSymbolTable();
-            for (const current of type.types) {
-                for (const prop of getPropertiesOfType(current)) {
-                    if (!members.has(prop.escapedName)) {
-                        const combinedProp = getPropertyOfUnionOrIntersectionType(type, prop.escapedName);
-                        if (combinedProp) {
-                            members.set(prop.escapedName, combinedProp);
+        for (const _ of iteratePropertiesOfUnionOrIntersectionType(type, /*doYield*/ false)) {
+            Debug.fail("Iterator should have been empty.");
+        }
+        return type.resolvedProperties!;
+    }
+
+    function* iteratePropertiesOfUnionOrIntersectionType(type: UnionOrIntersectionType, doYield = true) {
+        if (type.resolvedProperties) {
+            if (doYield) {
+                yield* type.resolvedProperties;
+            }
+            return;
+        }
+
+        if (doYield && type.partiallyResolvedProperties) {
+            yield* type.partiallyResolvedProperties;
+        }
+
+        if (!pushTypeResolution(type, TypeSystemPropertyName.ResolvedProperties)) {

```

Comment:
The fact that we didn't have this before is... spooky. Not sure what to even do here.

Reply:
We probably didn't bother, since, by construction, you can only have a union at the top level, followed by an intersection, and then none of either "deeper", so it's impossible for the actual structure of the type to recur. (And `this` types are too complicated to worry about most of the time.) Now, denormalized unions/intersections also make that less true, but we don't use denormalized types for anything other than type printback, so that's probably why it was usually fine.